compiler/optimizer/VPHandlers.cpp

```diff
@@ -2150,6 +2150,57 @@ TR::Node *constrainIaload(OMR::ValuePropagation *vp, TR::Node *node)
 
          node->setIsNonNull(true);
          }
+
+      else if (base
+               && base->getKnownObject()
+               && !symRef->isUnresolved() && symRef->getSymbol()->isRecognizedKnownObjectShadow()
+               && !vp->comp()->compileRelocatableCode())
+         // Base is known object, and field is recognized known object shadow,
+         // hence can be treated as known object.
+         {
+         TR::VMAccessCriticalSection constrainIaloadCriticalSection(vp->comp(),
+               TR::VMAccessCriticalSection::tryToAcquireVMAccess);
+         if (constrainIaloadCriticalSection.hasVMAccess())
+            {
+            TR::KnownObjectTable *knot = vp->comp()->getOrCreateKnownObjectTable();
+            uintptrj_t baseObject = knot->getPointer(base->getKnownObject()->getIndex());
+            // Check if baseObject is actually the recognized field Class
+            TR_OpaqueClassBlock *baseObjectClazz = TR::Compiler->cls.objectClass(vp->comp(), baseObject);
+            int32_t baseObjClassNameLength;
+            char *baseObjClassName = TR::Compiler->cls.classNameChars(vp->comp(), baseObjectClazz, baseObjClassNameLength);
+            int32_t recognizedClassNameLength;
+            const char *recognizedClassName = symRef->getSymbol()->owningClassNameCharsForRecognizedField(recognizedClassNameLength);
+            if (recognizedClassName && (baseObjClassNameLength == recognizedClassNameLength)
+                  && !strncmp(baseObjClassName, recognizedClassName, recognizedClassNameLength))
```

Comment:
This name-based test will work as long as the type can be uniquely identified by name (in Java that means: has a name that is only allowed to be defined by the bootstrap loader), and as long as we don't want to read the field from an instance of a subclass. If possible, I think it would be better to get the `TR_OpaqueClassBlock*` for the class containing the recognized field, and see whether `isInstanceOf(baseObjectClazz, recognizedFieldDefiningClazz, true, true) == TR_yes`

Reply:
Actually I thought you might suggest this. The recognized fields defining class we currently have are all basic bootloader classes so I thought this literal compare of the class name would work. Anyway, added isInstanceOf check.

Reply:
I think it would make more sense to pass `objectTypeIsFixed=true`, but that doesn't affect whether we get `TR_yes`